File: message/MessageCS/Message.cs
using System.IO;

namespace MessageCS
{
    /// <summary>
    /// 基于VarInt和ZigZag编码的消息
    /// </summary>
    public abstract class Message : Bean
    {
        public int Id { get; }

        public long Sn { get; set; }


        protected Message(int id)
        {
            Id = id;
        }

        public abstract Message Create();


        public override void Encode(Buffer buffer)
        {
            buffer.WriteInt(Id);
            buffer.WriteLong(Sn);
        }

        public override void Decode(Buffer buffer)
        {
            var id = buffer.ReadInt();
            if (id != Id)
            {
                throw new IOException("消息ID不匹配,目标值：" + Id + "，实际值：" + id);
            }

            Sn = buffer.ReadLong();
        }
    }
}

File: quan.java/message/src/main/java/quan/message/Message.java
package quan.message;

import java.io.IOException;

/**
 * 基于VarInt和ZigZag编码的消息<br/>
 * Created by quanchangnai on 2019/6/20.
 */
public abstract class Message extends Bean {

    /**
     * 消息ID
     */
    public abstract int getId();

    public abstract Message create();

    @Override
    public void encode(Buffer buffer) throws IOException {
        buffer.writeInt(getId());
    }

    @Override
    public void decode(Buffer buffer) throws IOException {
        int msgId = buffer.readInt();
        if (msgId != getId()) {
            throw new IOException(String.format("消息ID不匹配,期望值[%s],实际值[%s]", getId(), msgId));
        }
    }
}
